text/1522-conservative-impl-trait.md

```diff
@@ -534,6 +534,47 @@ But basically, without this feature certain things remain hard or impossible to
 in Rust, like returning a efficiently usable type parameterized by
 types private to a function body, for example an iterator adapter containing a closure.
 
+## Alternative to returnning unboxed closures
+
+Theoratically, if we are defining a function need to return an unboxed closure, it can always
+rewritten to a function that takes more parameters.
+
+For example, a fix point function takes a `Fn(T) -> R` and returns a `Fn(T) -> R` like this
+with the proposed `impl Trait` feature:
+
+```rust
+fn fix<T, R, F>(f: &F) -> impl Fn(T) -> R
+    where F: Fn(T) -> R
+{
+    f(|t| fix(f)(t))
+}
+...
+let closure = fix(func);
+```
+
+can be re-written as
+
+```rust
+fn fix<T, R, F>(f: F, t:T) -> R
+    where F: Fn(T) -> R
+{
+    f(|t|fix(f, t), t)
+}
+...
+let closure = |t|fix(func, t);
+```
```

Comment:
These aren't equivalent. In one the first function, you're returning a closure which can be evaluated with another argument at a later time. In the second, you're delaying running the function at all until it has been called with the last argument.

Consider, for example, the case where you add a `println!("Hello!");` to the beginning of each `fix` function:
```rust
// With `fix` version 1:
let closure = fix(func); // prints "Hello!"
closure(1); // Runs the closure, prints nothing
closure(1); // Runs the closure again, prints nothing

// With `fix` version 2:
let closure = |t| fix(func, t); // prints nothing
closure(1); // prints "Hello!"
closure(1); // prints "Hello!"
```